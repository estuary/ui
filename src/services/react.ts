import FullPageSpinner from 'components/fullPage/Spinner';
import { lazy } from 'react';
import { logRocketConsole, logRocketEvent } from './shared';
import { CustomEvents } from './types';

const LAZY_LOAD_FAILED_KEY = 'chunk_failed';

// https://mitchgavan.com/code-splitting-react-safely/
const setWithExpiry = (key: string, value: string, ttl: number) => {
    const item = {
        value,
        expiry: new Date().getTime() + ttl,
    };
    localStorage.setItem(key, JSON.stringify(item));
};

const getWithExpiry = (key: string) => {
    const itemString = window.localStorage.getItem(key);
    const item = itemString ? JSON.parse(itemString) : null;

    if (item === null) {
        return null;
    }

    // We have waited long enough to allow trying again so clearing out the key
    if (new Date().getTime() > item.expiry) {
        localStorage.removeItem(key);
        return null;
    }

    // Return value so we do not try reloading again
    return item.value;
};

const handledLazy = (factory: () => Promise<{ default: any }>) => {
    return lazy(() =>
        factory().catch((error) => {
            logRocketEvent(CustomEvents.LAZY_LOADING, 'failed');
            logRocketConsole('Component Failed Loading 1:', error);

            // Check if we're in an infinite loading loop before trying again
            if (!getWithExpiry(LAZY_LOAD_FAILED_KEY)) {
                setWithExpiry(LAZY_LOAD_FAILED_KEY, 'true', 10000);
                window.location.reload();
            }

            // Still make sure to return something so the app does not blow up
            return { default: FullPageSpinner };
        })
    );
};

export { handledLazy };
