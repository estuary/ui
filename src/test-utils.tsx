// https://testing-library.com/docs/react-testing-library/setup#custom-render

import { render, RenderOptions } from '@testing-library/react';
import { ReactElement, type FC } from 'react';
import { act } from 'react-dom/test-utils';
import { localStorageKey } from './auth';
import AppProviders from './context';

const AllTheProviders: FC = ({ children }) => {
    return <AppProviders>{children}</AppProviders>;
};

const customRender = (
    ui: ReactElement,
    options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

const loginAsUser = (userName: string = 'fakeUserName') => {
    act(() => {
        window.localStorage.setItem(localStorageKey, userName);
    });
};

export * from '@testing-library/react';
export { customRender as render, loginAsUser };
