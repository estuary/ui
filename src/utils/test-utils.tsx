/* eslint-disable @typescript-eslint/unbound-method */
// https://testing-library.com/docs/react-testing-library/setup#custom-render

import { render as rtlRender, RenderOptions } from '@testing-library/react';
import { AuthContext } from 'context/Auth';
import AppContent from 'context/Content';
import AppRouter from 'context/Router';
import AppTheme from 'context/Theme';
import { ReactElement } from 'react';
import { auth, AuthDetails } from 'services/auth';

// TODO - this does not really do anything because we hardcode
//   the username down below on the MockAuthProvider. However,
//   this will be needed in the future so leaving the work so far
//   so we can leverage this when wanting to test the "on load auth"
//   functionality
const loginAsUser = (username: string) => {
    const mockAuthDetails: AuthDetails = {
        session: {
            account_id: '',
            expires_at: '',
            token: '',
        },
        user: {
            created_at: '',
            display_name: username,
            email: `${username}@${username}`,
            id: '',
            name: username,
            unique_name: username,
            updated_at: '',
        },
    };

    auth.saveSession(mockAuthDetails.session);
    auth.saveUser(mockAuthDetails.user);
};

const logoutUser = () => {
    auth.removeAuthDetails();
};

const goTo = (route?: string, name?: string) => {
    window.history.pushState(
        {},
        name ? name : 'Test page',
        route ? route : '/'
    );
};

const waitForLoading = async () => {
    return Promise.resolve();
};

// Mocking AuthContext out basically skips all the loading screen and boostrapping
//  stuff that happens on load. This should be fine for unit testing as it
//  can help speed up tests a bit and makes things easier to deal with. This
//  does mean that we need to eventually write a basic test that handles the
//  account bootstrapping.
interface MockAuthProps {
    children: JSX.Element;
    username?: string;
}
const MockAuthProvider = ({ children, username }: MockAuthProps) => {
    const value = {
        login: jest.fn(() => {
            if (username) {
                loginAsUser(username);
            }
            return Promise.resolve();
        }),
        logout: jest.fn(() => {
            logoutUser();
            return Promise.resolve();
        }),
        user: username,
    };

    return (
        <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
    );
};

const customRender = async (
    ui: ReactElement,
    options: Omit<RenderOptions, 'wrapper'> & {
        route?: string;
        user?: string;
    }
) => {
    const { route, user } = options;

    goTo(route, 'Test Page');

    const view = rtlRender(
        <AppContent>
            <MockAuthProvider username={user}>
                <AppTheme>
                    <AppRouter>{ui}</AppRouter>
                </AppTheme>
            </MockAuthProvider>
        </AppContent>,
        {
            ...options,
        }
    );

    await waitForLoading();

    return view;
};

export * from '@testing-library/react';
export { goTo, loginAsUser, logoutUser, customRender, waitForLoading };
