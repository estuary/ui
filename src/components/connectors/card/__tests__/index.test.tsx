import { render } from '@testing-library/react';
import { renderOps } from 'test/test-utils';
import ConnectorCard from '..';

describe('Connector Card', () => {
    const logo = 'example.com/fake/logo.png';
    const details = 'this is a fake connector';
    const title = 'FakeDB';

    it('should display the connector details', async () => {
        const rendered = render(
            <ConnectorCard logo={logo} details={details} title={title} />,
            renderOps
        );
        expect(rendered.baseElement).toMatchSnapshot();
    });
});
