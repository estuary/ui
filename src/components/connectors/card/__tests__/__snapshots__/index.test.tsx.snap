// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Connector Card > should display the connector details 1`] = `
<body>
  <div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 MuiGrid-grid-md-4 MuiGrid-grid-lg-2 MuiGrid-grid-xl-2 css-13i4rnv-MuiGrid-root"
    >
      <button
        class="MuiButtonBase-root css-1un3d5i-MuiButtonBase-root"
        href=""
        tabindex="0"
        type="button"
      >
        <div
          class="MuiStack-root css-f03u9l-MuiStack-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <div
              class="MuiStack-root css-12atq1h-MuiStack-root"
            >
              <div
                class="MuiBox-root css-6ml8rd"
              >
                <div
                  class="MuiStack-root css-w004l2-MuiStack-root"
                >
                  <div
                    class="MuiBox-root css-0"
                  >
                    example.com/fake/logo.png
                  </div>
                </div>
              </div>
            </div>
            FakeDB
            this is a fake connector
          </div>
        </div>
      </button>
    </div>
  </div>
</body>
`;
